### PR TITLE
Allow new format for local_key

### LINKED ISSUE #409: Not a valid local key
**Describe the bug**
Iy9Ps(D;f*******, key for test (1024050068c63*******), is not a valid key.

**To Reproduce**
Reset the tuya device and use the new tuya local key from the tuya iot platform 

**Environment (please complete the following information):**
 - OS: Mac OS 
-  iOS version: 16.4.1
-  Homehubs: Homepod (new), Homepd mini, 2x Apple TV 4K2
 - Node Version: 18.16.0
 - Plugin Version: 2.0.1
-  Accessory Type: plug
Config file:
`

            "type": "Outlet",
            "name": "test",
            "id": "1024050068c63*******",
            "key": "Iy9Ps(D;f********",
            "manufacturer": "CSL"

`

**Additional context**
When using "old" key:

[5/11/2023, 1:42:12 PM] [homebridge-tuya] Discovered test (1024050068c63*******) identified as Outlet (3.3)
[Tuya] Changing ping gap for test to 9s
[Tuya DEBUG] reconnect called for test
[Tuya] Discovery ended.
[5/11/2023, 1:42:12 PM] [homebridge-tuya] Connected to test
[Tuya] Sending first query to test (3.3)
[Tuya] Odd message from test with command 10: \�J��q{?u���:f���G��#����~
[Tuya] Raw message from test (3.3) with command 10: 000055aa000000010000000a0000002c000000015ca24af9ae717b3f75f606a18e3a0c6697be8747ed1f7fbc23b092d510f1897e76eb92340000aa55
[Tuya DEBUG] decrementing this._connectionAttempts, currently 2
[Tuya DEBUG] decrementing this._connectionAttempts, currently 1

### LINKED ISSUE #410: Tuya local keys no longer work and have changed format?
**Checklist**
- [x ] I have read the common issues wiki page
- [ x] I have checked to make sure the plugin is up to date

**Describe the bug**

Extracted local keys from Tuya IOT platform no longer work, are case sensitive and include characters resulting in "[homebridge-tuya] ******************, key for Den Fan (xxxxxxxxxxxxxxxxxxxxxxxx), is not a valid key." error. 

**To Reproduce**
Steps to reproduce the behavior: Reset a connected cloud device and retrieve local key. Local key entry in config.js will not be accepted.

**Expected behavior**

Devices to function locally in HomeKit.

**Screenshots**

**Environment (please complete the following information):**
 - OS: macOS Ventura
-  iOS version:
-  Homehubs: HomePod mini
 - Node Version: v18.15.0
 - Plugin Version: v2.0.1
-  Accessory Type

**Additional context**

I had a Tuya switch drop off my network and had to reset, when retrieving new local keys I noticed they had changed format and no longer work in my integration.

### PR DESCRIPTION
**Pull request by justerror**
<s>Allow latin alphanumeric **and special character** for local_key</s>
Remove local_key check

Fixes #409 
Fixes #410

### CODE REVIEW

**Review by navarretemarce — APPROVED** [review]

File: index.js [hunk 1]
```diff
@@ -96,7 +96,7 @@ class TuyaLan {
             } catch(ex) {}
 
             //if (!/^[0-9a-f]+$/i.test(device.id)) return this.log.error('%s, id for %s, is not a valid id.', device.id, device.name || 'unnamed device');
-            if (!/^[0-9a-f]+$/i.test(device.key)) return this.log.error('%s, key for %s (%s), is not a valid key.', device.key.replace(/.{4}$/, '****'), device.name || 'unnamed device', device.id);
+            if (!/^[a-zA-Z0-9~`!@#$%^&*()+={}\[\];:\'\"<>.,\/\\\?-_]+$/i.test(device.key)) return this.log.error('%s, key for %s (%s), is not a valid key.', device.key.replace(/.{4}$/, '****'), device.name || 'unnamed device', device.id);
```
Comment (ascl00) [inline index.js:99]: I think the dash ('-') needs to be at the end else is interpreted as a range modifier (like the dash in a-z).

Comment (justerror) [reply index.js]: good catch, fixed

**Review by ascl00 — APPROVED** [review]
Looks good to me (FWIW).

**Review by guptaramm — APPROVED** [review]

**Comment by hosigi05** [comment]
Experience the new format not being allowed in the current version of this plugin too. Would be great this could be fixed in due course. Thanks @iRayanKhan

**Comment by SoCuul** [comment]
As I just discovered, this regex does not cover all local keys. One of my local keys failed the regex parsing, specifically it included a `|` character.

I modified the regex string a little, and it looks like this fixes the problem:
```regex
^[a-zA-Z0-9~`!@#$%^&*|()+={}\[\];:\'\"<>.,\/\\\?-_]+$
```

**Comment by ascl00** [comment]
At this point it might be just better to remove this check. It seems pretty much any ASCII character can be in the key, so continuing to extend the regex is not really helping.

**Comment by houmark** [comment]
+1 for merging this. I also encountered a new local key format today with both uppercase chars and special chars as well.

File: index.js [hunk 2]
```diff
@@ -96,7 +96,7 @@ class TuyaLan {
             } catch(ex) {}
 
             //if (!/^[0-9a-f]+$/i.test(device.id)) return this.log.error('%s, id for %s, is not a valid id.', device.id, device.name || 'unnamed device');
-            if (!/^[0-9a-f]+$/i.test(device.key)) return this.log.error('%s, key for %s (%s), is not a valid key.', device.key.replace(/.{4}$/, '****'), device.name || 'unnamed device', device.id);
+            if (!/^[a-zA-Z0-9~`!@#$%^&*()+={}\[\];:\'\"<>.,\/\\\?_-]+$/i.test(device.key)) return this.log.error('%s, key for %s (%s), is not a valid key.', device.key.replace(/.{4}$/, '****'), device.name || 'unnamed device', device.id);
             if (!{16:1, 24:1, 32: 1}[device.key.length]) return this.log.error('%s, key for %s (%s), doesn\'t have the expected length.', device.key.replace(/.{4}$/, '****'), device.name || 'unnamed device', device.id);
```
Comment (houmark) [inline index.js:100]: @justerror I would also remove this check now that we're at it, considering that the key could be any length and format now.

**Comment by justerror** [comment]
Removed all checks for local_key due to variability.

**Comment by hosigi05** [comment]
Getting this error when installing this update in my environment:

```
[9/15/2023, 9:35:56 AM] ====================
[9/15/2023, 9:35:56 AM] ERROR LOADING PLUGIN homebridge-tuya:
[9/15/2023, 9:35:56 AM] /var/lib/homebridge/node_modules/homebridge-tuya/lib/AirConditionerAccessory.js:432
            this.setMultiState({this.dpActive: true, this.dpRotationSpeed: this.convertRotationSpeedFromHomeKitToTuya(value)}, callback);

SyntaxError: Unexpected token '.'
    at internalCompileFunction (node:internal/vm:73:18)
    at wrapSafe (node:internal/modules/cjs/loader:1176:20)
    at Module._compile (node:internal/modules/cjs/loader:1218:27)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1308:10)
    at Module.load (node:internal/modules/cjs/loader:1117:32)
    at Function.Module._load (node:internal/modules/cjs/loader:958:12)
    at Module.require (node:internal/modules/cjs/loader:1141:19)
    at require (node:internal/modules/cjs/helpers:110:18)
    at Object.<anonymous> (/var/lib/homebridge/node_modules/homebridge-tuya/index.js:11:33)
    at Module._compile (node:internal/modules/cjs/loader:1254:14)
[9/15/2023, 9:35:56 AM] ====================
```

**Comment by lostdriveway** [comment]
Hi - new here and excuse my noobness - having this issue with my Arlec ceiling fan - how do you install this fix?

**Comment by SoCuul** [comment]
> Hi - new here and excuse my noobness - having this issue with my Arlec ceiling fan - how do you install this fix?

Hey! Since this isn't merged into the project yet, it'll make it harder to install.
What you can try is replacing a few lines in your currently installed version.
Keep in mind, all of these files will be located inside the `homebridge-tuya` folder inside `node_modules`.

## File 1: index.js
Replace this:
```js
if (!/^[0-9a-f]+$/i.test(device.key)) return this.log.error('%s, key for %s (%s), is not a valid key.', device.key.replace(/.{4}$/, '****'), device.name || 'unnamed device', device.id);
```
With this:
```js
if (!/^[a-zA-Z0-9~`!@#$%^&*|()+={}\[\];:\'\"<>.,\/\\\?-_]+$/i.test(device.key)) return this.log.error('%s, key for %s (%s), is not a valid key.', device.key.replace(/.{4}$/, '****'), device.name || 'unnamed device', device.id)
```

## File 2: cli-decode.js
Replace this:
```js
if (!/^[0-9a-f]+$/i.test(key)) {
```
With this:
```js
if (!/^[a-zA-Z0-9~`!@#$%^&*|()+={}\[\];:\'\"<>.,\/\\\?-_]+$/i.test(key)) {
```

---

This may not work (in the future), but it's the easiest solution for you right now. It definitely solved it for me :)

**Comment by lostdriveway** [comment]
Awesome thanks for that!  - now again excuse my basic skills

"Keep in mind, all of these files will be located inside the homebridge-tuya folder inside node_modules."

where is the node_modules folder usually found?

**Comment by SoCuul** [comment]
> Awesome thanks for that! - now again excuse my basic skills
> 
> "Keep in mind, all of these files will be located inside the homebridge-tuya folder inside node_modules."
> 
> where is the node_modules folder usually found?

If you're running homebridge on docker, you can find it by entering your container, by running these commands:
```sh
docker exec -it homebridge /bin/bash
cd homebridge/node_modules/homebridge-tuya
```

If you're running it without any complicated setup, just go into the `homebridge` folder, then you'll see `node_modules`, enter that and you'll see `tuya-homebridge`.

Hope that clarifies it a bit, but feel free to let me know if you need any help!

**Comment by lostdriveway** [comment]
Thanks I'll give it a crack when I get home from work this afternoon!

**Comment by lostdriveway** [comment]
Ok I can get this far- but no idea what the command is to edit the 2 files...sorry 

homebridge

**Comment by SoCuul** [comment]
> Ok I can get this far- but no idea what the command is to edit the 2 files...sorry
> 
> homebridge

From there, you can edit it with any text editor of your choice. I suggest using nano, which you can find tutorials for online!

**Comment by lostdriveway** [comment]
ok - did the first file - can't find the cli-decode.js file though?

**Comment by SoCuul** [comment]
> ok - did the first file - can't find the cli-decode.js file though?

`cli-decode.js` should be located in the `bin` folder

**Comment by michael-j-green** [comment]
Likely not the place to ask this, but is there a fork of this plugin that's actually being maintained?

This repo has 21 PR's, and 99 issues, and the maintainer appears to be MIA. I asked back in February if the maintainer was able to add another admin to the project to allow PR's to be merged, and releases to be made - crickets.

**Comment by iRayanKhan** [comment]
> Likely not the place to ask this, but is there a fork of this plugin that's actually being maintained?
> 
> 
> 
> This repo has 21 PR's, and 99 issues, and the maintainer appears to be MIA. I asked back in February if the maintainer was able to add another admin to the project to allow PR's to be merged, and releases to be made - crickets.


Is there a mystery admin who is willing to take care of it? Someone did message me on Twitter then I didn't hear further.

**Comment by lostdriveway** [comment]
Well I managed to find and make the requested changes to the code - no local key error message after reboot but it still can't find the device. 

I'll put it in the too hard basket for now until there's a better fix.

**Comment by SoCuul** [comment]
> > Likely not the place to ask this, but is there a fork of this plugin that's actually being maintained?
> > This repo has 21 PR's, and 99 issues, and the maintainer appears to be MIA. I asked back in February if the maintainer was able to add another admin to the project to allow PR's to be merged, and releases to be made - crickets.
> 
> Is there a mystery admin who is willing to take care of it? Someone did message me on Twitter then I didn't hear further.

@iRayanKhan is there a chance you could merge this into the master branch and publish it.
This plugin is still fully functional and works as intended, but this single local_key change pretty much makes the entire thing unusable.

**Comment by michael-j-green** [comment]
> > Likely not the place to ask this, but is there a fork of this plugin that's actually being maintained?
> > This repo has 21 PR's, and 99 issues, and the maintainer appears to be MIA. I asked back in February if the maintainer was able to add another admin to the project to allow PR's to be merged, and releases to be made - crickets.
> 
> Is there a mystery admin who is willing to take care of it? Someone did message me on Twitter then I didn't hear further.

Good question... I sadly don't have an answer for you. I'm not knowledgable enough with Homebridge or the inner workings of this plugin to take it on myself.

Perhaps one of the more active fork maintainers, or one of the more knowledgable users in this PR like @SoCuul (assuming they want to take it on)?

**Comment by SoCuul** [comment]
> > > Likely not the place to ask this, but is there a fork of this plugin that's actually being maintained?
> 
> > > This repo has 21 PR's, and 99 issues, and the maintainer appears to be MIA. I asked back in February if the maintainer was able to add another admin to the project to allow PR's to be merged, and releases to be made - crickets.
> 
> > 
> 
> > Is there a mystery admin who is willing to take care of it? Someone did message me on Twitter then I didn't hear further.
> 
> 
> 
> Good question... I sadly don't have an answer for you. I'm not knowledgable enough with Homebridge or the inner workings of this plugin to take it on myself.
> 
> 
> 
> Perhaps one of the more active fork maintainers, or one of the more knowledgable users in this PR like @SoCuul (assuming they want to take it on)?

I would be happy to help out, but while I'm fluent in typescript, I have little to no experience building a homebridge plug-in. I've gone over the documentation countless times, but I haven't fully got the grasp of it.

**Comment by lostdriveway** [comment]
Finally got it working thanks to some help in the discord

had to add "//" in front of line 99 on the index.js and also added a line into the config.json and it finally showed up in Homebridge

**Comment by michael-j-green** [comment]
> > > > Likely not the place to ask this, but is there a fork of this plugin that's actually being maintained?
> > 
> > 
> > > > This repo has 21 PR's, and 99 issues, and the maintainer appears to be MIA. I asked back in February if the maintainer was able to add another admin to the project to allow PR's to be merged, and releases to be made - crickets.
> > 
> > 
> > > 
> > 
> > 
> > > Is there a mystery admin who is willing to take care of it? Someone did message me on Twitter then I didn't hear further.
> > 
> > 
> > Good question... I sadly don't have an answer for you. I'm not knowledgable enough with Homebridge or the inner workings of this plugin to take it on myself.
> > Perhaps one of the more active fork maintainers, or one of the more knowledgable users in this PR like @SoCuul (assuming they want to take it on)?
> 
> I would be happy to help out, but while I'm fluent in typescript, I have little to no experience building a homebridge plug-in. I've gone over the documentation countless times, but I haven't fully got the grasp of it.

That's fair :) Sorry to call you out like that!

**Comment by ayushsharma82** [comment]
A working fork till then for easy install: https://github.com/ayushsharma82/homebridge-tuya-local

@iRayanKhan  Feel free to pull and merge.

**Comment by 05TEVE** [comment]
> > > > > Likely not the place to ask this, but is there a fork of this plugin that's actually being maintained?
> > > 
> > > 
> > > > > This repo has 21 PR's, and 99 issues, and the maintainer appears to be MIA. I asked back in February if the maintainer was able to add another admin to the project to allow PR's to be merged, and releases to be made - crickets.
> > > 
> > > 
> > > > 
> > > 
> > > 
> > > > Is there a mystery admin who is willing to take care of it? Someone did message me on Twitter then I didn't hear further.
> > > 
> > > 
> > > Good question... I sadly don't have an answer for you. I'm not knowledgable enough with Homebridge or the inner workings of this plugin to take it on myself.
> > > Perhaps one of the more active fork maintainers, or one of the more knowledgable users in this PR like @SoCuul (assuming they want to take it on)?
> > 
> > 
> > I would be happy to help out, but while I'm fluent in typescript, I have little to no experience building a homebridge plug-in. I've gone over the documentation countless times, but I haven't fully got the grasp of it.
> 
> That's fair :) Sorry to call you out like that!

I am happy to help out and try keep this plugin alive. I messaged @iRayanKhan a while back on twitter but never 

> > Likely not the place to ask this, but is there a fork of this plugin that's actually being maintained?
> > This repo has 21 PR's, and 99 issues, and the maintainer appears to be MIA. I asked back in February if the maintainer was able to add another admin to the project to allow PR's to be merged, and releases to be made - crickets.
> 
> Is there a mystery admin who is willing to take care of it? Someone did message me on Twitter then I didn't hear further.

That was me that reached out. I have sent a request to add you on Discord to discuss further.

**Comment by 05TEVE** [comment]
> A working fork till then for easy install: https://github.com/ayushsharma82/homebridge-tuya-local
> 
> @iRayanKhan Feel free to pull and merge.



> A working fork till then for easy install: https://github.com/ayushsharma82/homebridge-tuya-local
> 
> @iRayanKhan Feel free to pull and merge.

Hey @ayushsharma82, I have spoken with @iRayanKhan and got access to merge PRs into the repository with the intention to start fixing some of the issues listed. Looking at the changes you have made i am thinking it might make sense to start with merging your fork back and publishing a new version?

**Review by 05TEVE — APPROVED** [review]
Changes look good to me.

**Comment by 05TEVE** [comment]
Changes have been published up to NPM as version 3.1.0-beta.2. Should be available for testing on homebridge.

**Pull request merged**